### PR TITLE
🐛 media-sound/audacity: fix various issues

### PR DESCRIPTION
**Pull request by Jamim**
_Hello everyone,_

The `flac` USE flag was mistakenly used instead of the `opus` USE flag for configuring.

Also, **Audacity** `3.4.2` [introduced](https://github.com/audacity/audacity/pull/5370) a brand new `USE_OPUSFILE` option which must be set properly.

These changes:
  - fix the `opus` USE flag
  - add proper `REQUIRED_USE`
  - replace `media-sound/mpg123` with `media-sound/mpg123-base`
  - move `dev-libs/rapidjson` from `RDEPEND` to `DEPEND`
  - fix missing `#include <limits>` for `2.4.2-r3`
  - limit `media-video/ffmpeg` to `<5` for `2.4.2-r3`
  - add `x11-base/xorg-proto` to `DEPEND`
  - add some ASM to `BDEPEND`
  - add `audacity-3.4.2-audiocom-std-string.patch` to `PATCHES` for `3.3.3-r1`
  - add `audacity-2.4.2-fix-libflac-undefined-references.patch` to `PATCHES` for `2.4.2-r4`
  - reset `KEYWORDS` for `2.4.2-r4` and `3.3.3-r1`

<details>
<summary>Closes:</summary>

- https://bugs.gentoo.org/741969
- https://bugs.gentoo.org/884747
- https://bugs.gentoo.org/910723
- https://bugs.gentoo.org/922595
- https://bugs.gentoo.org/927659

</details>

_Best regards!_

### CODE REVIEW

**Comment by Jamim** [comment]
This PR supersedes #35895.

File: media-sound/audacity/audacity-3.4.2.ebuild
Comment (Jamim) [reply media-sound/audacity/audacity-3.4.2.ebuild:106]: > Because it needs a bump, I would consider solving `media-sound/mpg123-base` in this PR (albeit it can be as a separate commit), else that would mean another bump.
> 
> And, I spotted another one: looks like `dev-libs/rapidjson` is headers only. I think meaning it would be a DEPEND only, and not trigger a rebuild on slot change.

I've addressed these imperfections in 05d838a28b1c2516741f7ff25b2c4e89984e76e8.

File: media-sound/audacity/audacity-3.4.2-r1.ebuild
Comment (Jamim) [reply media-sound/audacity/audacity-3.4.2-r1.ebuild]: `mod-ogg` is a bit misleading name. Effectively, it's a `mod-vorbis` under the hood, so I've added `vorbis? ( ogg )` to the `REQUIRED_USE`.

**Comment by Jamim** [comment]
_Hello @mehw,_
Would you mind reviewing this PR?
_Thanks in advance!_

**Comment by mehw** [comment]
Hello @Jamim,

Do you have any suggestion about the tools to use in an LXC container (rif. https://github.com/gentoo/gentoo/pull/35887#issuecomment-2035841408)?
```
app-portage/iwdevtools
dev-util/pkgcheck
dev-util/pkgdev
```
This is what I read to prepare the LXC container for Audacity:
- https://wiki.gentoo.org/wiki/Test_environment#Using_LXC_containers_for_ebuild_testing
- https://wiki.gentoo.org/wiki/LXC

LXC container: `gentoo 20240402_16:07 amd64 openrc`.

File: media-sound/audacity/audacity-3.4.2-r1.ebuild
Comment (mehw) [inline media-sound/audacity/audacity-3.4.2-r1.ebuild:117]: `CMakeLists.txt` lists `NASM` as a requirement.

```
The ASM_NASM compiler identification is unknown
Didn't find assembler

CMakeLists.txt:280: include( CMakeDetermineASM_NASMCompiler )

diff -Nuar a/audacity-3.4.2-r1.ebuild b/audacity-3.4.2-r1.ebuild
--- a/audacity-3.4.2-r1.ebuild	2024-04-07 20:56:04.464000000 +0000
+++ b/audacity-3.4.2-r1.ebuild	2024-04-07 23:59:20.715888551 +0000
@@ -114,6 +114,7 @@
 	dev-libs/rapidjson
 	test? ( <dev-cpp/catch-3:0 )"
 BDEPEND="
+	dev-lang/nasm
 	sys-devel/gettext
 	virtual/pkgconfig
 "
```

**Review by mehw — CHANGES_REQUESTED** [review]

**Comment by mehw** [comment]
@Jamim @cjmayo I'll gather the result of my testing here: https://github.com/mehw/gentoo-ebuild-testing

Interesting enough, some required dependencies might be currently missing.

This lead my compilation to a failure in an LXC container:
```
emerge =media-sound/audacity-3.4.2-r1
emerge --unmerge media-sound/audacity
emerge --depclean
rm /var/cache/binpkgs/media-sound/audacity/audacity-3.4.2-r1-*
emerge -k =media-sound/audacity-3.4.2-r1
```

The last `emerge -k` merged less packages than the first one (without `-k`), and the compilation failed on `xproto` and `NASM`.

The fix is to add `x11-base/xorg-proto` as DEPEND and `dev-lang/nasm` as BDEPEND.

File: media-sound/audacity/audacity-3.4.2-r1.ebuild
Comment (Jamim) [reply media-sound/audacity/audacity-3.4.2-r1.ebuild:117]: I've figured out that Yasm is fine, too.
```
# grep -i asm *
audacity-3.4.2:20240403-010706.log:-- The ASM_NASM compiler identification is NASM
audacity-3.4.2:20240403-010706.log:-- Found assembler: /usr/bin/nasm
audacity-3.4.2:20240408-012932.log:-- The ASM_NASM compiler identification is YASM
audacity-3.4.2:20240408-012932.log:-- Found assembler: /usr/bin/yasm
```

**Comment by Jamim** [comment]
Thank you for the great job, @mehw! 🙇🏼
I've updated the PR.

**Comment by dlan17** [comment]
hi, guys, is there anyone from proaudio team can do the proxy commit? otherwise I can help, just let me know when the PR is ready (good to go)..

File: media-sound/audacity/audacity-3.4.2-r1.ebuild
Comment (mehw) [reply media-sound/audacity/audacity-3.4.2-r1.ebuild:118]: Thanks ;)

**Comment by fordfrog** [comment]
> hi, guys, is there anyone from proaudio team can do the proxy commit? otherwise I can help, just let me know when the PR is ready (good to go)..

@dlan17 feel free to take care of it, thanks :-)

**Comment by mehw** [comment]
@Jamim @dlan17 `media-sound/audacity-3.4.2-r1` passed all the tests that I've done, even compiling with all the use flags enabled and `FEATURES="test"`.

I'll take a look in the logs to see if I can find any mismatch with the build requirements.

I'm going to try `media-sound/audacity-3.3.3-r1` in the next hours.

**Comment by dlan17** [comment]
> @Jamim @dlan17 `media-sound/audacity-3.4.2-r1` passed all the tests that I've done, even compiling with all the use flags enabled and `FEATURES="test"`.
> 
I've noticed there is an error, have you fixed it already?
https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/results/PR_35903/head_05d838a28b1c2516741f7ff25b2c4e89984e76e8/media-sound/audacity/audacity-3.4.2-r1/2/audacity-3.4.2-r1-2.build

> I'll take a look in the logs to see if I can find any mismatch with the build requirements.
> 
Ok

> I'm going to try `media-sound/audacity-3.3.3-r1` in the next hours.
Ok, keep me posted then.. and let me know explicitly if it's fine to merge, thanks

**Comment by mehw** [comment]
> I've noticed there is an error, have you fixed it already?
> https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/results/PR_35903/head_05d838a28b1c2516741f7ff25b2c4e89984e76e8/media-sound/audacity/audacity-3.4.2-r1/2/audacity-3.4.2-r1-2.build

There was a missing [DEPEND="x11-base/xorg-proto"](https://github.com/gentoo/gentoo/pull/35903#discussion_r1555113535) and a missing [BDEPEND="dev-lang/nasm"](https://github.com/gentoo/gentoo/pull/35903#discussion_r1555115426): [fixed](https://github.com/gentoo/gentoo/pull/35903#issuecomment-2041832131) by @Jamim in the latest https://github.com/gentoo/gentoo/commit/512e17d21024862b7e2f58025fe72354863ee345 (appending `DEPEND="x11-base/xorg-proto"`, and `BDEPEND="|| ( dev-lang/nasm dev-lang/yasm )"`.

> Ok, keep me posted then.. and let me know explicitly if it's fine to merge, thanks

Sure, I'll let you know.

**Comment by Jamim** [comment]
_Hello @dlan,_

> I've noticed there is an error, have you fixed it already?

That failure relates to the older revision of the changes.
It was fixed by adding `x11-base/xorg-proto` to `DEPEND` as it was suggested [here](https://github.com/gentoo/gentoo/pull/35903#issuecomment-2041668723) by @mehw.

Please note that @mehw's [README](https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/README.md) contains headers, such as [PR 35903 HEAD 05d838a](https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/README.md#pr-35903-head-05d838a) which indicates that the following reports are related to #35903 and 05d838a28b1c2516741f7ff25b2c4e89984e76e8.

**Comment by mehw** [comment]
Hello @Jamim,

It seems that `media-sound/audacity-3.3.3-r1` requires `audacity-3.4.2-audiocom-std-string.patch` from [PR#35414](https://github.com/gentoo/gentoo/pull/35414), rif. [bug#920363](https://bugs.gentoo.org/920363), which was not included into `media-sound/audacity-3.3.3` ebuild after merging.

```
diff -Nuar a/audacity-3.3.3-r1.ebuild b/audacity-3.3.3-r1.ebuild
--- a/audacity-3.3.3-r1.ebuild	2024-04-08 08:46:13.038864921 +0000
+++ b/audacity-3.3.3-r1.ebuild	2024-04-11 23:30:59.522440714 +0000
@@ -140,6 +140,9 @@
 
 	# Allows running tests without conan
 	"${FILESDIR}/${PN}-3.3.3-remove-conan-test-dependency.patch"
+
+	# #920363
+	"${FILESDIR}/audacity-3.4.2-audiocom-std-string.patch"
 )
 
 src_prepare() {
```

**Comment by Jamim** [comment]
> It seems that `media-sound/audacity-3.3.3-r1` requires `audacity-3.4.2-audiocom-std-string.patch` from [PR#35414](https://github.com/gentoo/gentoo/pull/35414), rif. [bug#920363](https://bugs.gentoo.org/920363), which was not included into `media-sound/audacity-3.3.3` ebuild after merging.

I've addressed it in 911aff12f7fe07a527a7faa15c78156b1a15b8f7. Thank you, @mehw!

**Comment by Jamim** [comment]
Taking into account how broken the USE flags at the moment, I believe it's better to merge these fixes now.
If there are any other issues, then we can address them in follow-up PRs.
@mehw, @dlan17 what do you think?

**Comment by dlan17** [comment]
> Taking into account how broken the USE flags are at the moment, I believe it's better to merge these fixes now. If there are any other issues, then we can address them in follow-up PRs. @mehw, @dlan17 what do you think?

basically, I'd leave @mehw to judge, I'm fine to merge if it's not a severe problem (security related, or a regression?), doing in a progressive improvement way would be good, no need always waiting

**Comment by mehw** [comment]
> > It seems that `media-sound/audacity-3.3.3-r1` requires `audacity-3.4.2-audiocom-std-string.patch` from [PR#35414](https://github.com/gentoo/gentoo/pull/35414), rif. [bug#920363](https://bugs.gentoo.org/920363), which was not included into `media-sound/audacity-3.3.3` ebuild after merging.
> 
> I've addressed it in [911aff1](https://github.com/gentoo/gentoo/commit/911aff12f7fe07a527a7faa15c78156b1a15b8f7). Thank you, @mehw!

Great!

> > Taking into account how broken the USE flags are at the moment, I believe it's better to merge these fixes now. If there are any other issues, then we can address them in follow-up PRs. @mehw, @dlan17 what do you think?
> 
> basically, I'd leave @mehw to judge, I'm fine to merge if it's not a severe problem (security related, or a regression?), doing in a progressive improvement way would be good, no need always waiting

@Jamim @dlan17 I'm fine with merging, I ran a bunch of tests for `3.3.3-r1` and `3.4.2-r1`, all passed for [911aff1](https://github.com/gentoo/gentoo/commit/911aff12f7fe07a527a7faa15c78156b1a15b8f7).

I can run the same tests for `2.4.2-r3` even after merging, and explore the logs thoroughly for all versions later on.

By the way, upstream released a [3.5.0-beta-3](https://github.com/audacity/audacity/releases/tag/Audacity-3.5.0-beta-3), any one of you is interested in a WIP?

Thanks.

**Review by mehw — APPROVED** [review]

**Comment by dlan17** [comment]
I've got these warnings when run "pkgcheck scan --commits $(git merge-base HEAD origin/master)..HEAD", 

  VariableShadowed: version 3.3.3-r1: variable 'REQUIRED_USE' may be shadowed, on lines: 41, 115
  DroppedKeywords: version 3.4.2-r1: ppc
  RdependChange: version 2.4.2-r3: RDEPEND modified without revbump

 VariableShadowed:
   REQUIRED_USE="mad? ( !mpg123 )" is dropped, is it intentional? as I didn't see this restriction in 3.4.2
 DroppedKeywords:
   I'd leave it to you and no need actions in this PR, you can file a bug at b.g.o to request ARCH testing team to handle it
RdependChange
  strictly we should give it a revision bump, as it will fix rdeps in some cases (like binhost)

**Comment by dlan17** [comment]
> @Jamim @dlan17 I'm fine with merging, I ran a bunch of tests for `3.3.3-r1` and `3.4.2-r1`, all passed for [911aff1](https://github.com/gentoo/gentoo/commit/911aff12f7fe07a527a7faa15c78156b1a15b8f7).
> 
> I can run the same tests for `2.4.2-r3` even after merging, and explore the logs thoroughly for all versions later on.
> 
sounds good to me

> By the way, upstream released a [3.5.0-beta-3](https://github.com/audacity/audacity/releases/tag/Audacity-3.5.0-beta-3), any one of you is interested in a WIP?
> 
No, I'd leave this up to you, I'm here merely to do some last step checking & help to merge this PR..

**Comment by mehw** [comment]
@Jamim could you please take a look at [audacity-2.4.2-r3/2/files/audacity-2.4.2-fix-libflac-undefined-references.patch](https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/results/PR_35903/head_911aff12f7fe07a527a7faa15c78156b1a15b8f7/media-sound/audacity/audacity-2.4.2-r3/2/files/audacity-2.4.2-fix-libflac-undefined-references.patch) that closes [bug#741969](https://bugs.gentoo.org/741969)?

```
diff -Nuar a/audacity-2.4.2-r3.ebuild b/audacity-2.4.2-r3.ebuild
--- a/audacity-2.4.2-r3.ebuild	2024-04-08 08:46:13.038864921 +0000
+++ b/audacity-2.4.2-r3.ebuild	2024-04-14 05:10:20.119372126 +0000
@@ -68,6 +68,7 @@
 	"${FILESDIR}/${P}-add-missing-include-limits.patch"
 	"${FILESDIR}/${P}-add-missing-include-portaudio.patch"
 	"${FILESDIR}/${P}-disable-ccache.patch"
+	"${FILESDIR}/${P}-fix-libflac-undefined-references.patch"
 )
 
 src_prepare() {
```

Thanks.

IMHO, due to new patches: `audacity-2.4.2-r3` -> `audacity-2.4.2-r4` and set it as unstable `KEYWORDS="~amd64 ~arm64 ~ppc ~ppc64 ~riscv ~x86"`.

If appropriate, also unstabilize `audacity-3.3.3-r1`.

What do you think?

> I've got these warnings when run "pkgcheck scan --commits $(git merge-base HEAD origin/master)..HEAD",
> 
> VariableShadowed: version 3.3.3-r1: variable 'REQUIRED_USE' may be shadowed, on lines: 41, 115 DroppedKeywords: version 3.4.2-r1: ppc RdependChange: version 2.4.2-r3: RDEPEND modified without revbump

@dlan17 thanks for verifying, I'll take a look.

> VariableShadowed: for REQUIRED_USE change in version 3.3.3-r1, it doesn't take effect (as be shadowed, see warning msg) REQUIRED_USE="mad? ( !mpg123 )" is dropped in 3.4.2, is it intentional? as I didn't see this restriction DroppedKeywords: I'd leave it to you and no need actions in this PR, you can file a bug at b.g.o to request ARCH testing team to handle it RdependChange strictly we should give it a revision bump, as it will fix rdeps in some cases (like binhost)

About the missing `REQUIRED_USE="mad? ( !mpg123 )"` in `3.4.2`, libmad support had been dropped:
- https://github.com/audacity/audacity/pull/4917#issuecomment-1636719217
- https://github.com/audacity/audacity/pull/4668

**Comment by dlan17** [comment]
> @Jamim could you please take a look at [audacity-2.4.2-r3/2/files/audacity-2.4.2-fix-libflac-undefined-references.patch](https://github.com/mehw/gentoo-ebuild-testing/blob/media-sound/audacity/results/PR_35903/head_911aff12f7fe07a527a7faa15c78156b1a15b8f7/media-sound/audacity/audacity-2.4.2-r3/2/files/audacity-2.4.2-fix-libflac-undefined-references.patch) that closes [bug#741969](https://bugs.gentoo.org/741969) and [bug#884747](https://bugs.gentoo.org/884747)?
> 
> ```
> diff -Nuar a/audacity-2.4.2-r3.ebuild b/audacity-2.4.2-r3.ebuild
> --- a/audacity-2.4.2-r3.ebuild	2024-04-08 08:46:13.038864921 +0000
> +++ b/audacity-2.4.2-r3.ebuild	2024-04-14 05:10:20.119372126 +0000
> @@ -68,6 +68,7 @@
>  	"${FILESDIR}/${P}-add-missing-include-limits.patch"
>  	"${FILESDIR}/${P}-add-missing-include-portaudio.patch"
>  	"${FILESDIR}/${P}-disable-ccache.patch"
> +	"${FILESDIR}/${P}-fix-libflac-undefined-references.patch"
>  )
>  
>  src_prepare() {
> ```
> 
> Thanks.
> 
> IMHO, due to new patches: `audacity-2.4.2-r3` -> `audacity-2.4.2-r4` and set it as unstable `KEYWORDS="~amd64 ~arm64 ~ppc ~ppc64 ~riscv ~x86"`.
> 
> If appropriate, also unstabilize `audacity-3.3.3-r1`.
> 
> What do you think?
> 
Yes, sounds good if you want to drop them to unstable keywords, and this is the normal way how we handle new patch,
in rare case you can request a fast stabilization (security bugs related), or bump the stable keywords if the
change is trivial or really safe?

> > I've got these warnings when run "pkgcheck scan --commits $(git merge-base HEAD origin/master)..HEAD",
> > VariableShadowed: version 3.3.3-r1: variable 'REQUIRED_USE' may be shadowed, on lines: 41, 115 DroppedKeywords: version 3.4.2-r1: ppc RdependChange: version 2.4.2-r3: RDEPEND modified without revbump
> 
> @dlan17 thanks for verifying, I'll take a look.
> 
> > VariableShadowed: for REQUIRED_USE change in version 3.3.3-r1, it doesn't take effect (as be shadowed, see warning msg) REQUIRED_USE="mad? ( !mpg123 )" is dropped in 3.4.2, is it intentional? as I didn't see this restriction DroppedKeywords: I'd leave it to you and no need actions in this PR, you can file a bug at b.g.o to request ARCH testing team to handle it RdependChange strictly we should give it a revision bump, as it will fix rdeps in some cases (like binhost)
> 
> About the missing `REQUIRED_USE="mad? ( !mpg123 )"` in `3.4.2`, libmad support had been dropped:
> 
Ok, thanks for explanation

>     * [Add missing Findlibmad.cmake file to resolve the system installed libmad audacity/audacity#4917 (comment)](https://github.com/audacity/audacity/pull/4917#issuecomment-1636719217)
> 
>     * [New export UI audacity/audacity#4668](https://github.com/audacity/audacity/pull/4668)

**Comment by Jamim** [comment]
_Hello @dlan17 and @mehw,_
Hope I've addressed all the recently mentioned issues in 649075c51b9d1a0b8342021bcedc249383a8ce84.

**Comment by dlan17** [comment]
> _Hello @dlan17 and @mehw,_ Hope I've addressed all the recently mentioned issues in [649075c](https://github.com/gentoo/gentoo/commit/649075c51b9d1a0b8342021bcedc249383a8ce84).

thanks, I've pushed with only change that version 2.4.2-r3 has been kept, as some ARCHs only has stable keywords in this version, later you may want to start filing stabilization bug for it..

**Comment by mehw** [comment]
@Jamim @dlan17 Nice! Thank you ;)

**Comment by Jamim** [comment]
We finally did it! 🎉
Thank you for your help @mehw, @dlan17, and @cjmayo! 🙇🏼

**Pull request closed without merging**
